### PR TITLE
[#169647021] added registration page 3

### PR DESCRIPTION
**Pull request by d-al-ibm**
added translations for page 3
moved token setting and user profile api call to DefaultContainer (as token and profile info are used in every child component of DefaultContainer)
added registration step 3
added integration with get documents api
moved loading page to default container (as it is used only after login)
added new env variable to api env file

### CODE REVIEW

File: src/components/DefaultContainer/DefaultContainer.tsx [hunk 1]
```diff
@@ -70,40 +84,117 @@ export const DefaultContainer = () => {
...
-      {...props}
-      onGetUserProfile={handleGetUserProfile}
-      spidMail={userProfile.email}
-      onWorkMailSet={handleWorkMailSet}
-      toggleAddMailModal={toggleAddMailModal}
+  /**
+   * Given a cookie key `cookieName`, returns the value of
+   * the cookie or empty string, if the key is not found.
+   */
+  function getCookie(cookieName: string): string {
```
Comment (gunzip) [inline src/components/DefaultContainer/DefaultContainer.tsx]: please move this function in a utility package or - better - use some library like https://www.npmjs.com/package/react-cookie

File: src/components/DefaultContainer/DefaultContainer.tsx [hunk 2]
```diff
@@ -70,40 +84,117 @@ export const DefaultContainer = () => {
...
+
+  /**
+   * Set token in token context
+   */
+  useEffect(() => {
+    const token = getCookie("sessionToken");
+    // if getCookie returns an empty string and the user is not in pre login page (where token still has to be set),
+    // it means the token has expired and user browser deleted it -> redirect user to login, otherwise set tokenContext.token
+    if (
+      NonEmptyString.is(token) ||
```
Comment (gunzip) [inline src/components/DefaultContainer/DefaultContainer.tsx]: please name conditions ie. `const isTokenExpired = NonEmptyString.is(token) || ...`

File: src/components/DefaultContainer/DefaultContainer.tsx [hunk 3]
```diff
@@ -70,40 +84,117 @@ export const DefaultContainer = () => {
...
+    // TODO: when available, redirect to logout page (tracked in story https://www.pivotaltracker.com/story/show/169033467)
+    else {
+      props.history.push("/home");
+    }
+  }, [location.pathname]);
+
+  useEffect(() => {
+    // make api call only after onMount because token is string in any case, no longer undefined, and only if userProfile is not set and user is not on spid login page
+    if (
+      NonEmptyString.is(tokenContext.token) &&
```
Comment (gunzip) [inline src/components/DefaultContainer/DefaultContainer.tsx]: same as above

File: src/components/DefaultContainer/DefaultContainer.tsx [hunk 4]
```diff
@@ -70,40 +84,117 @@ export const DefaultContainer = () => {
...
+      !NonEmptyString.is(userProfile.given_name) &&
+      location.pathname !== "/spid-login"
+    ) {
+      const url =
+        customWindow._env_.IO_ONBOARDING_PA_API_HOST +
+        ":" +
+        customWindow._env_.IO_ONBOARDING_PA_API_PORT +
+        "/profile";
+      fetch(url, {
```
Comment (gunzip) [inline src/components/DefaultContainer/DefaultContainer.tsx]: please add a todo comment linked to pivotal story to refactor these calls with io.utils generated code

File: src/components/Registration/RegistrationStepThree/RegistrationStepThree.tsx [hunk 5]
```diff
@@ -1,8 +1,221 @@
...
+  const customWindow = (window as unknown) as ICustomWindow;
+
+  const urlDomainPort =
+    customWindow._env_.IO_ONBOARDING_PA_API_HOST +
+    ":" +
+    customWindow._env_.IO_ONBOARDING_PA_API_PORT;
+
+  const tokenContext = useContext(TokenContext);
+
+  const showFile = (blob: Blob, documentName: string) => {
```
Comment (gunzip) [inline src/components/Registration/RegistrationStepThree/RegistrationStepThree.tsx]: please move this method into some utility module

Comment (gunzip) [reply src/components/Registration/RegistrationStepThree/RegistrationStepThree.tsx]: better: conider using something like http://danml.com/download.html

File: src/components/Registration/RegistrationStepThree/RegistrationStepThree.tsx [hunk 6]
```diff
@@ -1,8 +1,221 @@
...
+      // For Firefox it is necessary to delay revoking the ObjectURL
+      window.URL.revokeObjectURL(data);
+    }, 100);
+  };
+
+  const downloadDocument = () => (_: MouseEvent) => {
+    const url =
+      urlDomainPort +
+      `/organizations/${props.ipaCode}/documents/${props.documentName}`;
+    fetch(url, {
```
Comment (gunzip) [inline src/components/Registration/RegistrationStepThree/RegistrationStepThree.tsx]: add comment

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This PR changes a total of 4473 LOCs, that is more than a reasonable size of 250. Consider splitting the pull request into smaller ones.</td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#169647021](https://www.pivotaltracker.com/story/show/169647021): Aggiungere pagina caricamento creazione documenti


New dependencies added: <a href='https://www.npmjs.com/package/file-saver'><code>file-saver</code></a>, <a href='https://www.npmjs.com/package/react-cookie'><code>react-cookie</code></a> and <a href='https://www.npmjs.com/package/@types/file-saver'><code>&#64;types/file-saver</code></a>.


<h2><a href="https://github.com/eligrey/FileSaver.js#readme">file-saver</a></h2>
<p>Author: Eli Grey</p>
<p>Description: An HTML5 saveAs() FileSaver implementation</p>
<p>Homepage: <a href="https://github.com/eligrey/FileSaver.js#readme">https://github.com/eligrey/FileSaver.js#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>5 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>2</td></tr><tr><td>Releases</td><td>16</td></tr><tr><td>Keywords</td><td>filesaver, saveas and blob</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

If you need to save really large files bigger then the blob's size limitation or don't have
enough RAM, then have a look at the more advanced [StreamSaver.js][7]
that can save data directly to the hard drive asynchronously with the power of the new streams API. That will have
support for progress, cancelation and knowing when it's done writing

FileSaver.js
============

FileSaver.js is the solution to saving files on the client-side, and is perfect for
web apps that generates files on the client, However if the file is coming from the
server we recommend you to first try to use [Content-Disposition][8] attachment response header as it has more cross-browser compatiblity.

Looking for `canvas.toBlob()` for saving canvases? Check out
[canvas-toBlob.js][2] for a cross-browser implementation.

Supported Browsers
------------------

| Browser        | Constructs as | Filenames    | Max Blob Size | Dependencies |
| -------------- | ------------- | ------------ | ------------- | ------------ |
| Firefox 20+    | Blob          | Yes          | 800 MiB       | None         |
| Firefox < 20   | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |
| Chrome         | Blob          | Yes          | [2GB][3]      | None         |
| Chrome for Android | Blob      | Yes          | [RAM/5][3]    | None         |
| Edge           | Blob          | Yes          | ?             | None         |
| IE 10+         | Blob          | Yes          | 600 MiB       | None         |
| Opera 15+      | Blob          | Yes          | 500 MiB       | None         |
| Opera < 15     | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |
| Safari 6.1+*   | Blob          | No           | ?             | None         |
| Safari < 6     | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |
| Safari 10.1+   | Blob          | Yes          | n/a           | None         |

Feature detection is possible:

```js
try {
    var isFileSaverSupported = !!new Blob;
} catch (e) {}
```

### IE < 10

It is possible to save text files in IE < 10 without Flash-based polyfills.
See [ChenWenBrian and koffsyrup's `saveTextAs()`](https://github.com/koffsyrup/FileSaver.js#examples) for more details.

### Safari 6.1+

Blobs may be opened instead of saved sometimes—you may have to direct your Safari users to manually
press <kbd>⌘</kbd>+<kbd>S</kbd> to save the file after it is opened. Using the `application/octet-stream` MIME type to force downloads [can cause issues in Safari](https://github.com/eligrey/FileSaver.js/issues/12#issuecomment-47247096).

### iOS

saveAs must be run within a user interaction event such as onTouchDown or onClick; setTimeout will prevent saveAs from triggering. Due to restrictions in iOS saveAs opens in a new window instead of downloading, if you want this fixed please [tell Apple how this WebKit bug is affecting you](https://bugs.webkit.org/show_bug.cgi?id=167341).

Syntax
------
### Import `saveAs()` from file-saver
```js
import { saveAs } from 'file-saver';
```

```js
FileSaver saveAs(Blob/File/Url, optional DOMString filename, optional Object { autoBom })
```

Pass `{ autoBom: true }` if you want FileSaver.js to automatically provide Unicode text encoding hints (see: [byte order mark](https://en.wikipedia.org/wiki/Byte_order_mark)). Note that this is only done if your blob type has `charset=utf-8` set.

Examples
--------

### Saving text using `require()`
```js
var FileSaver = require('file-saver');
var blob = new Blob(["Hello, world!"], {type: "text/plain;charset=utf-8"});
FileSaver.saveAs(blob, "hello world.txt");
```

### Saving text

```js
var blob = new Blob(["Hello, world!"], {type: "text/plain;charset=utf-8"});
FileSaver.saveAs(blob, "hello world.txt");
```

### Saving URLs

```js
FileSaver.saveAs("https://httpbin.org/image", "image.jpg");
```
Using URLs within the same origin will just use `a[download]`.
Otherwise, it will first check if it supports cors header with a synchronous head request.
If it does, it will download the data and save using blob URLs. 
If not, it will try to download it using `a[download]`.

The standard W3C File API [`Blob`][4] interface is not available in all browsers.
[Blob.js][5] is a cross-browser `Blob` implementation that solves this.

### Saving a canvas
```js
var canvas = document.getElementById("my-canvas");
canvas.toBlob(function(blob) {
    saveAs(blob, "pretty image.png");
});
```

Note: The standard HTML5 `canvas.toBlob()` method is not available in all browsers.
[canvas-toBlob.js][6] is a cross-browser `canvas.toBlob()` that polyfills this.

### Saving File

You can save a File constructor without specifying a filename. If the
file itself already contains a name, there is a hand full of ways to get a file
instance (from storage, file input, new constructor, clipboard event). 
If you still want to change the name, then you can change it in the 2nd argument.

```js
// Note: Ie and Edge don't support the new File constructor,
// so it's better to construct blobs and use saveAs(blob, filename)
var file = new File(["Hello, world!"], "hello world.txt", {type: "text/plain;charset=utf-8"});
FileSaver.saveAs(file);
```



Tracking image

  [1]: http://eligrey.com/demos/FileSaver.js/
  [2]: https://github.com/eligrey/canvas-toBlob.js
  [3]: https://bugs.chromium.org/p/chromium/issues/detail?id=375297#c107
  [4]: https://developer.mozilla.org/en-US/docs/DOM/Blob
  [5]: https://github.com/eligrey/Blob.js
  [6]: https://github.com/eligrey/canvas-toBlob.js
  [7]: https://github.com/jimmywarting/StreamSaver.js
  [8]: https://github.com/eligrey/FileSaver.js/wiki/Saving-a-remote-file#using-http-header

Installation
------------------

```bash
# Basic Node.JS installation
npm install file-saver --save
bower install file-saver
```

Additionally, TypeScript definitions can be installed via:

```bash
# Additional typescript definitions
npm install @types/file-saver --save-dev
```


</details>




<h2><a href="https://github.com/reactivestack/cookies/tree/master/packages/react-cookie/#readme">react-cookie</a></h2>
<p>Author: Benoit Tremblay</p>
<p>Description: Universal cookies for React</p>
<p>Homepage: <a href="https://github.com/reactivestack/cookies/tree/master/packages/react-cookie/#readme">https://github.com/reactivestack/cookies/tree/master/packages/react-cookie/#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 4 years ago</td></tr><tr><td>Last Updated</td><td>4 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>63</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/@types/hoist-non-react-statics'><code>&#64;types/hoist-non-react-statics</code></a>, <a href='https://www.npmjs.com/package/hoist-non-react-statics'><code>hoist-non-react-statics</code></a> and <a href='https://www.npmjs.com/package/universal-cookie'><code>universal-cookie</code></a></td></tr><tr><td>Keywords</td><td>universal, isomophic, cookie and react</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

<h3 align="center">
  <a href="packages/react-cookie">react-cookie</a>
</h3>

<p align="center">
  Universal cookies for <a href="https://facebook.github.io/react">React</a><br />
  <a href="https://badge.fury.io/js/react-cookie"></a>
</p>

[Build Status](https://travis-ci.org/reactivestack/cookies)
<br />
[Sauce Test Status](https://saucelabs.com/u/coookies)

## Integrations

- [`universal-cookie`](https://www.npmjs.com/package/universal-cookie) - Universal cookies for JavaScript
- [`universal-cookie-express`](https://www.npmjs.com/package/universal-cookie-express) - Hook cookies get/set on Express for server-rendering

## Minimum requirement

### react-cookie @ v3.0+

- React.js >= 16.3.0 (new context API + forward ref)

### react-cookie @ v0.0-v2.2

- React.js >= 15

## Getting started

`npm install react-cookie`

or in the browser (global variable `ReactCookie`):

```html
<script
  crossorigin
  src="https://unpkg.com/react@16/umd/react.production.js"
></script>
<script
  crossorigin
  src="https://unpkg.com/universal-cookie@3/umd/universalCookie.min.js"
></script>
<script
  crossorigin
  src="https://unpkg.com/react-cookie@3/umd/reactCookie.min.js"
></script>
```

## `<CookiesProvider />`

Set the user cookies

On the server, the `cookies` props must be set using `req.universalCookies` or `new Cookie(cookieHeader)`

## `useCookies([dependencies])`

Access and modify cookies using React hooks.

```jsx
const [cookies, setCookie, removeCookie] = useCookies(['cookie-name']);
```

**React hooks are available starting from React 16.8**

### `dependencies` (optional)

Let you optionally specify a list of cookie names your component depend on or that should trigger a re-render. If unspecified, it will render on every cookie change.

### `cookies`

Javascript object with all your cookies. The key is the cookie name.

### `setCookie(name, value, [options])`

Set a cookie value

- name (string): cookie name
- value (string|object): save the value and stringify the object if needed
- options (object): Support all the cookie options from RFC 6265
  - path (string): cookie path, use `/` as the path if you want your cookie to be accessible on all pages
  - expires (Date): absolute expiration date for the cookie
  - maxAge (number): relative max age of the cookie from when the client receives it in second
  - domain (string): domain for the cookie (sub.domain.com or .allsubdomains.com)
  - secure (boolean): Is only accessible through HTTPS?
  - httpOnly (boolean): Is only the server can access the cookie?
  - sameSite (boolean|lax|strict): Strict or Lax enforcement

### `removeCookie(name, [options])`

Remove a cookie

- name (string): cookie name
- options (object): Support all the cookie options from RFC 6265
  - path (string): cookie path, use `/` as the path if you want your cookie to be accessible on all pages
  - expires (Date): absolute expiration date for the cookie
  - maxAge (number): relative max age of the cookie from when the client receives it in second
  - domain (string): domain for the cookie (sub.domain.com or .allsubdomains.com)
  - secure (boolean): Is only accessible through HTTPS?
  - httpOnly (boolean): Is only the server can access the cookie?
  - sameSite (boolean|lax|strict): Strict or Lax enforcement

## `withCookies(Component)`

Give access to your cookies anywhere. Add the following props to your component:

- cookies: Cookies instance allowing you to get, set and remove cookies.
- allCookies: All your current cookies in an object.

Your original static properties will be hoisted on the returned component. You can also access the original component by using the `WrappedComponent` static property. Example:

```jsx
function MyComponent() {
  return null;
}
const NewComponent = withCookies(MyComponent);
NewComponent.WrappedComponent === MyComponent;
```

## Cookies

### `get(name, [options])`

Get a cookie value

- name (string): cookie name
- options (object):
  - doNotParse (boolean): do not convert the cookie into an object no matter what

### `getAll([options])`

Get all cookies

- options (object):
  - doNotParse (boolean): do not convert the cookie into an object no matter what

### `set(name, value, [options])`

Set a cookie value

- name (string): cookie name
- value (string|object): save the value and stringify the object if needed
- options (object): Support all the cookie options from RFC 6265
  - path (string): cookie path, use `/` as the path if you want your cookie to be accessible on all pages
  - expires (Date): absolute expiration date for the cookie
  - maxAge (number): relative max age of the cookie from when the client receives it in second
  - domain (string): domain for the cookie (sub.domain.com or .allsubdomains.com)
  - secure (boolean): Is only accessible through HTTPS?
  - httpOnly (boolean): Is only the server can access the cookie?
  - sameSite (boolean|lax|strict): Strict or Lax enforcement

### `remove(name, [options])`

Remove a cookie

- name (string): cookie name
- options (object): Support all the cookie options from RFC 6265
  - path (string): cookie path, use `/` as the path if you want your cookie to be accessible on all pages
  - expires (Date): absolute expiration date for the cookie
  - maxAge (number): relative max age of the cookie from when the client receives it in second
  - domain (string): domain for the cookie (sub.domain.com or .allsubdomains.com)
  - secure (boolean): Is only accessible through HTTPS?
  - httpOnly (boolean): Is only the server can access the cookie?
  - sameSite (boolean|lax|strict): Strict or Lax enforcement

## Simple Example with React hooks

```js
// Root.jsx
import React from 'react';
import App from './App';
import { CookiesProvider } from 'react-cookie';

export default function Root() {
  return (
    <CookiesProvider>
      <App />
    </CookiesProvider>
  );
}
```

```js
// App.jsx
import React from 'react';
import { useCookies } from 'react-cookie';

import NameForm from './NameForm';

function App() {
  const [cookies, setCookie] = useCookies(['name']);

  function onChange(newName) {
    setCookie('name', newName, { path: '/' });
  }

  return (
    <div>
      <NameForm name={cookies.name} onChange={onChange} />
      {cookies.name && <h1>Hello {cookies.name}!</h1>}
    </div>
  );
}

export default App;
```

## Simple Example with Higher-Order Component

```js
// Root.jsx
import React from 'react';
import App from './App';
import { CookiesProvider } from 'react-cookie';

export default function Root() {
  return (
    <CookiesProvider>
      <App />
    </CookiesProvider>
  );
}
```

```js
// App.jsx
import React, { Component } from 'react';
import { instanceOf } from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';

import NameForm from './NameForm';

class App extends Component {
  static propTypes = {
    cookies: instanceOf(Cookies).isRequired
  };

  constructor(props) {
    super(props);

    const { cookies } = props;
    this.state = {
      name: cookies.get('name') || 'Ben'
    };
  }

  handleNameChange(name) {
    const { cookies } = this.props;

    cookies.set('name', name, { path: '/' });
    this.setState({ name });
  }

  render() {
    const { name } = this.state;

    return (
      <div>
        <NameForm name={name} onChange={this.handleNameChange.bind(this)} />
        {this.state.name && <h1>Hello {this.state.name}!</h1>}
      </div>
    );
  }
}

export default withCookies(App);
```

## Server-Rendering Example

```js
// src/components/App.js
import React from 'react';
import { useCookies } from 'react-cookie';

import NameForm from './NameForm';

function App() {
  const [cookies, setCookie] = useCookies(['name']);

  function onChange(newName) {
    setCookie('name', newName, { path: '/' });
  }

  return (
    <div>
      <NameForm name={cookies.name} onChange={onChange} />
      {cookies.name && <h1>Hello {cookies.name}!</h1>}
    </div>
  );
}

export default App;
```

```js
// src/server.js
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { CookiesProvider } from 'react-cookie';

import Html from './components/Html';
import App from './components/App';

export default function middleware(req, res) {
  const markup = ReactDOMServer.renderToString(
    <CookiesProvider cookies={req.universalCookies}>
      <App />
    </CookiesProvider>
  );

  const html = ReactDOMServer.renderToStaticMarkup(<Html markup={markup} />);

  res.send('<!DOCTYPE html>' + html);
}
```

```js
// src/client.js
import React from 'react';
import ReactDOM from 'react-dom';
import { CookiesProvider } from 'react-cookie';

import App from './components/App';

const appEl = document.getElementById('main-app');

ReactDOM.render(
  <CookiesProvider>
    <App />
  </CookiesProvider>,
  appEl
);
```

```js
// server.js
require('@babel/register');

const express = require('express');
const serverMiddleware = require('./src/server').default;
const cookiesMiddleware = require('universal-cookie-express');

const app = express();

app
  .use('/assets', express.static('dist'))
  .use(cookiesMiddleware())
  .use(serverMiddleware);

app.listen(8080, function() {
  console.log('Listening on 8080...');
});
```


</details>




<h2><a href="http://npmjs.com/package/@types/file-saver">&#64;types/file-saver</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for FileSaver.js</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/file-saver">http://npmjs.com/package/@types/file-saver</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 3 years ago</td></tr><tr><td>Last Updated</td><td>9 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>6</td></tr><tr><td>Direct Dependencies</td><td></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/file-saver`

# Summary
This package contains type definitions for FileSaver.js ( https://github.com/eligrey/FileSaver.js/ ).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/file-saver

Additional Details
 * Last updated: Tue, 21 May 2019 19:56:14 GMT
 * Dependencies: none
 * Global values: none

# Credits
These definitions were written by Cyril Schumacher <https://github.com/cyrilschumacher>, Daniel Roth <https://github.com/DaIgeb>, Chris Barr <https://github.com/chrismbarr>, HitkoDev <https://github.com/HitkoDev>.


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5069460b09d861ae0cfc06325437d26ac92f53f7
</p>

**Pull request merged**